modules/core/include/opencv2/core/utils/lock.hpp

```diff
@@ -0,0 +1,119 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#ifndef OPENCV_UTILS_LOCK_HPP
+#define OPENCV_UTILS_LOCK_HPP
+
+namespace cv { namespace utils {
+
+
+/** @brief A simple scoped lock (RAII-style locking for exclusive/write access).
+ *
+ * Emulate std::lock_guard (C++11), partially std::unique_lock (C++11),
+ */
+template <class _Mutex>
+class lock_guard {
+public:
+    typedef _Mutex Mutex;
+
+    explicit inline lock_guard(Mutex &m) : mutex_(&m) { mutex_->lock(); }
+
+    inline ~lock_guard() { if (mutex_) mutex_->unlock(); }
+
+    inline void release()
+    {
+        CV_DbgAssert(mutex_);
+        mutex_->unlock();
+        mutex_ = NULL;
+    }
+
+private:
+    Mutex* mutex_;
+
+private:
+    lock_guard(const lock_guard&); // disabled
+    lock_guard& operator=(const lock_guard&); // disabled
+};
+
+
+/** @brief A shared scoped lock (RAII-style locking for shared/reader access).
+ *
+ * Emulate boost::shared_lock_guard, subset of std::shared_lock (C++14),
+ */
+template <class _Mutex>
+class shared_lock_guard {
+public:
+    typedef _Mutex Mutex;
+
+    explicit inline shared_lock_guard(Mutex &m) : mutex_(&m) { mutex_->lock_sharable(); }
+
+    inline ~shared_lock_guard() { if (mutex_) mutex_->unlock_sharable(); }
+
+    inline void release()
+    {
+        CV_DbgAssert(mutex_);
+        mutex_->unlock_sharable();
+        mutex_ = NULL;
+    }
+
+protected:
+    Mutex* mutex_;
+
+private:
+    shared_lock_guard(const shared_lock_guard&); // disabled
+    shared_lock_guard& operator=(const shared_lock_guard&); // disabled
+};
+
+
+/** @brief An optional simple scoped lock (RAII-style locking for exclusive/write access).
+ *
+ * Doesn't lock if mutex pointer is NULL.
+ *
+ * @sa lock_guard
+ */
+template <class _Mutex>
+class optional_lock_guard {
+public:
+    typedef _Mutex Mutex;
+
+    explicit inline optional_lock_guard(Mutex* m) : mutex_(m) { if (mutex_) mutex_->lock(); }
+
+    inline ~optional_lock_guard() { if (mutex_) mutex_->unlock(); }
```

Comment:
can we use mixed case name for the class?

Reply:
This is not OpenCV specific name and not a part of external public API.

Name is selected to be consistent with declarations above: `lock_guard`, `shared_lock_guard` (both have alternative implementation in modern C++ / boost library under the same names).

File is renamed to "lock.private.hpp".